packages/firestore/src/core/query.ts

```diff
@@ -211,55 +245,48 @@ export function isCollectionGroupQuery(query: Query): boolean {
 }
 
 /**
- * Returns the implicit order by constraint that is used to execute the Query,
- * which can be different from the order by constraints the user provided (e.g.
- * the SDK and backend always orders by `__name__`).
+ * Returns the order by constraint that is used to execute the Query, in which the implicit order by
+ * constraint can be different from the order by constraints the user provided (e.g. the SDK and
+ * backend always orders by `__name__`).
  */
 export function queryOrderBy(query: Query): OrderBy[] {
   const queryImpl = debugCast(query, QueryImpl);
   if (queryImpl.memoizedOrderBy === null) {
     queryImpl.memoizedOrderBy = [];
+    const fieldsIncluded = new Set<string>();
 
-    const inequalityField = getInequalityFilterField(queryImpl);
-    const firstOrderByField = getFirstOrderByField(queryImpl);
-    if (inequalityField !== null && firstOrderByField === null) {
-      // In order to implicitly add key ordering, we must also add the
-      // inequality filter field for it to be a valid query.
-      // Note that the default inequality field and key ordering is ascending.
-      if (!inequalityField.isKeyField()) {
-        queryImpl.memoizedOrderBy.push(new OrderBy(inequalityField));
+    // Any explicit order by fields should be added as is.
+    for (const orderBy of queryImpl.explicitOrderBy) {
+      queryImpl.memoizedOrderBy.push(orderBy);
+      fieldsIncluded.add(orderBy.field.canonicalString());
+    }
+
+    // The order of the implicit ordering always matches the last explicit order by.
+    const lastDirection =
+      queryImpl.explicitOrderBy.length > 0
+        ? queryImpl.explicitOrderBy[queryImpl.explicitOrderBy.length - 1].dir
+        : Direction.ASCENDING;
+
+    // Any inequality field not explicitly ordered should be implicitly ordered in a lexicographical
+    // order. When there are multiple inequality filters on the same field, the field should be added
+    // only once.
+    const inequalityFields = getInequalityFilterFields(queryImpl);
```

Comment:
if `getInequalityFilterFields` returns a sorted set, you won't need the `fieldsIncluded` in this function.

Reply:
🤔  Still need it, cause if a field is explicitly ordered, it shouldn't be implicitly ordered again. Checking membership of a set (: fieldsIncluded  ≈ explicitOrderBy ) is better than checking it on array(: queryImpl.explicitOrderBy)

Reply:
plus, checking the membership in FieldFilter[] might need an extra function to compare items.

Reply:
Ah, yes!